src/main/java/org/killbill/billing/plugin/adyen/client/payment/service/AdyenPaymentServiceProviderPort.java

```diff
@@ -156,70 +156,54 @@ public PurchaseResult authoriseKlarnaPayment(final String merchantAccount,
                                                  final PaymentData paymentData,
                                                  final UserData userData) {
         PaymentsRequest klarnaRequest = adyenRequestFactory.createKlarnaPayment(merchantAccount, paymentData, userData);
-        PurchaseResult purchaseResult = null;
-        try {
-            PaymentsResponse response = adyenCheckoutApiClient.createPayment(klarnaRequest);
-            purchaseResult = extractKlarnaResponse(response);
-        }
-        catch (ApiException ex) {
-            new PurchaseResult(PaymentServiceProviderResult.valueOf("error"),
-                    null,
-                    null,
-                    ex.getMessage(),
-                    null,
-                    klarnaRequest.getReference(), null);
-        } catch (IOException ex) {
-            new PurchaseResult(PaymentServiceProviderResult.valueOf("error"),
-                    null,
-                    null,
-                    ex.getMessage(),
-                    null,
-                    klarnaRequest.getReference(), null);
+        final AdyenCallResult<PaymentsResponse> adyenCallResult = adyenCheckoutApiClient.createPayment(klarnaRequest);
+        if (!adyenCallResult.receivedWellFormedResponse()) {
+            return handleKlarnaAuthoriseError(adyenCallResult, paymentData);
         }
 
-        return purchaseResult;
+        return handleKlarnaAuthoriseResponse(adyenCallResult, merchantAccount);
     }
 
     public PurchaseResult completeKlarnaPaymentAuth(final String merchantAccount,
                                                  final PaymentData paymentData,
                                                  final UserData userData) {
         PaymentsDetailsRequest detailsRequest = adyenRequestFactory.completeKlarnaPayment(merchantAccount, paymentData, userData);
-        PurchaseResult purchaseResult = null;
-        try {
-            PaymentsResponse response = adyenCheckoutApiClient.paymentDetails(detailsRequest);
-            purchaseResult = extractKlarnaResponse(response);
+        final AdyenCallResult<PaymentsResponse> adyenCallResult = adyenCheckoutApiClient.paymentDetails(detailsRequest);
+        if (!adyenCallResult.receivedWellFormedResponse()) {
+            return handleKlarnaAuthoriseError(adyenCallResult, paymentData);
         }
-        catch (ApiException ex) {
+        return handleKlarnaAuthoriseResponse(adyenCallResult, merchantAccount);
+    }
+
+    private PurchaseResult handleKlarnaAuthoriseError(AdyenCallResult<PaymentsResponse> callResult, final PaymentData paymentData) {
```

Comment:
change the name. it is confusing as it is used for both autocomplete and authorize

Reply:
Both functions are part of Authorise hence the error handling is applicable for both cases. However, I have refactored for code for authoriseKlarnaPayment  and completeKlarnaPaymentAuth functions to use same name, authoriseKlarnaPayment with additional flag to indicate complete auth call.

Reply:
I think checkout API is generic. We should have generic Response handling.
Do we really need Klarna specific error and response handling ?